Open VideoCapture from data stream

dkurt opened the pull request:
### Pull Request Readiness Checklist

Add VideoCapture option to read a raw binary video data from `std::streambuf`.

There are multiple motivations:
1. Avoid disk file creation in case of video already in memory (received by network or from database).
2. Streaming mode. Frames decoding starts during sequential file transfer by chunks.

Suppoted backends:
* FFmpeg
* MSMF (no streaming mode)

Supporter interfaces:
* C++ (std::streambuf)
* Python (io.BufferedIOBase)

resolves https://github.com/opencv/opencv/issues/24400

- [x] test h264
- [x]  test IP camera like approach with no metadata but key frame only?
- [x] C API plugin

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

opencv-alalek reviewed (commented):
Plugin support is necessary - it is mandatory for FFmpeg support on Windows.

opencv-alalek commented on modules/videoio/src/plugin_capture_api.hpp:
```diff
@@ -121,6 +121,27 @@ struct OpenCV_VideoIO_Capture_Plugin_API_v1_1_api_entries
+
+    @param filename File name or NULL to use camera_index or buffer instead
+    @param camera_index Camera index (used if filename == NULL and buffer == NULL)
+    @param buffer Memory buffer pointer (used if filename == NULL)
```
Do we really need to handle all 3 parameters for the source in a single method of API?
As we keep the previous version of the method, then we don't need to duplicate its functionality.

dkurt replied:
Sorry foe delay. Changed to a method which accepts only buffer.

Wanted to add GStreamer backend but couldn't make it work for now.

dkurt commented:
Discussed on OpenCV weekly sync about potential improvements:
1. On current PR state, do not use `std::vector` but allow user pass just pointer/buffer_size.
2. Consider alternative API with a kind of streaming by chunks. Might be more efficient with GPU decoders.

gdippolito commented:
Hello, what is left to complete this work? Would anyone be able to complete this work by any chance?

gdippolito commented:
Hey @dkurt thanks for your work on this PR. Do you think this is ready for some testing?

dkurt commented:
@gdippolito, thanks! Yes, it's ready and any feedback is welcome.

opencv-alalek commented on modules/videoio/src/plugin_capture_api.hpp:
```diff
@@ -121,6 +121,23 @@ struct OpenCV_VideoIO_Capture_Plugin_API_v1_1_api_entries
+    @note API-CALL 9, API-Version == 2
+     */
+    CvResult (CV_API_CALL *Capture_open_buffer)(
+        std::streambuf& buffer,
```
We should not pass C++ objects through low level plugin API. They are not usable in general.

Different C++ ABIs are possible for OpenCV and plugin parts:
- MSVS vs MinGW
- GCC/Clang with different C++ standards
- libc++ (GCC) / libstdc++ (Clang)

dkurt replied:
Thanks! That was my concern. Is that a good idea to replace with functions pointers for required methods like `seek` and `read`?

dkurt commented on modules/videoio/misc/python/pyopencv_videoio.hpp:
```diff
@@ -1,6 +1,8 @@
 #ifdef HAVE_OPENCV_VIDEOIO
+#include "../../src/cap_interface.hpp"
```
Is that ok to have relative import path? I don't want to move `CvStream` to public API

mshabunin replied:
This approach might not work with standalone python bindings build or some other special configuration.

As an alternative you can make it "technically" public by moving class declaration to a `.private.hpp` header like we do with [avi container](https://github.com/opencv/opencv/blob/4.x/modules/videoio/include/opencv2/videoio/container_avi.private.hpp) - symbols will be exported, but header will not be installed.

dkurt replied:
Thanks. If that's not critical, I moved to `utils.private.hpp`. We can avoid this by small duplication of the struct `CvStream : public std::streambuf` but in bindings.

dkurt commented:
- [ ] 
  ```
  [  FAILED  ] 3 tests, listed below:
  [  FAILED  ] videoio/buffer_capture.read/12, where GetParam() = ("mpg", MSMF)
  [  FAILED  ] videoio/buffer_capture.read/20, where GetParam() = ("h264", MSMF)
  [  FAILED  ] videoio/buffer_capture.read/24, where GetParam() = ("h265", MSMF)
  ```

dkurt commented on modules/videoio/src/plugin_capture_api.hpp:
```diff
@@ -121,6 +121,25 @@ struct OpenCV_VideoIO_Capture_Plugin_API_v1_1_api_entries
+{
+    /** @brief Open video capture from buffer with parameters
+
+    @param buffer Memory buffer
```
- [ ] Update docstring with new arguments

dkurt commented:
@asmorkalov, @opencv-alalek, I added test with container extensions ("h264", "h265", "mjpg.avi"). Also fixed plugin API to accept just two methods - seek and read.

The only remaining part is MSMF code - for now it reads entire buffer but not but chunks. But I think it can be improved separately.

asmorkalov commented:
@dkurt Please rebase and fix conflict before merge.

dkurt commented:
@asmorkalov, I try final experiments with MSMF for today and if no luck, I'll rebase for a final review.

opencv-alalek reviewed (commented):
Need to provide own simplified callback interface (`cv::IReadStream`) with read+seek operations.

dkurt commented on modules/videoio/src/cap_ffmpeg.cpp:
```diff
@@ -312,6 +332,46 @@ CvResult CV_API_CALL cv_capture_open_with_params(
+        return CV_ERROR_FAIL;
+    *handle = NULL;
+    if (buffer.sgetc() == EOF)
+        return CV_ERROR_FAIL;
```
`sgetc` does not modify sequence position. I also worried about that but checked again

dkurt replied:
Removed completely

dkurt commented on modules/videoio/src/backend_plugin.cpp:
```diff
@@ -704,6 +731,25 @@ Ptr<IVideoCapture> PluginBackend::createCapture(const std::string &filename, con
     return Ptr<IVideoCapture>();
 }
 
+Ptr<IVideoCapture> PluginBackend::createCapture(std::streambuf& source, const VideoCaptureParameters& params) const
```
Thanks for the review. Added a simplified interface, but it’s private for internal conversion.

opencv-alalek commented on modules/videoio/include/opencv2/videoio.hpp:
```diff
@@ -798,6 +798,14 @@ class CV_EXPORTS_W VideoCapture
     CV_WRAP explicit VideoCapture(int index, int apiPreference, const std::vector<int>& params);
 
+    /** @overload
+    @brief Opens a video file stored in memory buffer.
```
> in memory buffer

Opens a video using data stream.

opencv-alalek commented on modules/videoio/include/opencv2/videoio/utils.private.hpp:
```diff
@@ -10,6 +10,83 @@
+};
+
+
+class ReadStreamCallback : public IReadStream
```
`ReadStreamPluginProvider`

Usually `Callback` means abstract interface description. We already have the `IReadStream` for that.

opencv-alalek commented on modules/videoio/misc/python/pyopencv_videoio.hpp:
```diff
@@ -31,4 +31,76 @@ template<> bool pyopencv_to(PyObject* obj, cv::VideoCapture& stream, const ArgIn
+        PyObject* res = PyObject_CallMethodObjArgs(ioBase, PyString_FromString("read"), size, NULL);
+        char* src = PyBytes_AsString(res);
+        size_t len = static_cast<size_t>(PyBytes_Size(res));
+        std::memcpy(buf, src, len);
```
Check `len < n`

opencv-alalek commented on modules/videoio/misc/python/pyopencv_videoio.hpp:
```diff
@@ -31,4 +31,76 @@ template<> bool pyopencv_to(PyObject* obj, cv::VideoCapture& stream, const ArgIn
+    }
+
+private:
+    PyObject* obj;
```
Who calls `Py_DECREF` ?

dkurt replied:
Added `Py_INCREF` as well

opencv-alalek commented on modules/videoio/src/cap.cpp:
```diff
@@ -228,6 +235,128 @@ bool VideoCapture::open(const String& filename, int apiPreference, const std::ve
+            {
+                try
+                {
+                    icap = backend->createCapture(StreambufReadStream::create(stream), parameters);
```
Who performs `seek(0, 0)` between trials?

Perhaps we should not allow `CAP_ANY` in this API.

dkurt replied:
Good point, there is no explicit seek. Added a check that `apiPref != CAP_ANY`

opencv-alalek commented on modules/videoio/include/opencv2/videoio/utils.private.hpp:
```diff
@@ -10,6 +10,83 @@
+    }
+
+private:
+    std::streambuf& stream;
```
There is no guarantee for lifetime of this.

Perhaps we need to pass `std::shared_ptr<std::streambuf>` in public API.

dkurt replied:
Agree, I think it can simplify other bindings addition in future.

opencv-alalek commented on modules/videoio/include/opencv2/videoio/utils.private.hpp:
```diff
@@ -10,6 +10,83 @@
+
+    IReadStream* clone() override
+    {
+        return new StreambufReadStream(stream);
```
Looks useless because it could not (and should not) create independent objects (at least seek pointer is shared).

See below comment about lifetime.

dkurt replied:
Removed `clone` method completely in last commit. As I understand, `struct CvCapture_FFMPEG` cannot hold a copy of `Ptr<IReadStream>` because there is no default constructor to allocate it: https://github.com/opencv/opencv/blob/b42075f3e2c7eca4a12e55b32cfc95fa90982a21/modules/videoio/src/cap_ffmpeg_impl.hpp#L3296

So I used `CvCapture_FFMPEG_proxy` to increase a counter

opencv-alalek commented on modules/videoio/src/videoio_registry.cpp:
```diff
@@ -45,9 +45,9 @@ namespace {
 #endif
 
-#define DECLARE_STATIC_BACKEND(cap, name, mode, createCaptureFile, createCaptureCamera, createWriter) \
+#define DECLARE_STATIC_BACKEND(cap, name, mode, createCaptureFile, createCaptureCamera, createCaptureBuffer, createWriter) \
```
Please add new macro instead of massive refactoring:

`DECLARE_STATIC_BACKEND_WITH_STREAM_SUPPORT`

opencv-alalek commented on modules/videoio/src/videoio_registry.hpp:
```diff
@@ -14,6 +14,7 @@ namespace cv
 enum BackendMode {
     MODE_CAPTURE_BY_INDEX    = 1 << 0,           //!< device index
     MODE_CAPTURE_BY_FILENAME = 1 << 1,           //!< filename or device path (v4l2)
+    MODE_CAPTURE_BY_STREAM   = 1 << 2,           //!< capture from memory buffer
```
need to update comment too.

opencv-alalek commented on modules/videoio/src/videoio_registry.hpp:
```diff
@@ -38,6 +39,7 @@ namespace videoio_registry {
 
 std::vector<VideoBackendInfo> getAvailableBackends_CaptureByIndex();
 std::vector<VideoBackendInfo> getAvailableBackends_CaptureByFilename();
+std::vector<VideoBackendInfo> getAvailableBackends_CaptureByBuffer();
```
`getAvailableBackends_CaptureByStream`

dkurt commented on modules/videoio/include/opencv2/videoio/registry.hpp:
```diff
@@ -35,6 +35,9 @@ CV_EXPORTS_W std::vector<VideoCaptureAPIs> getCameraBackends();
 CV_EXPORTS_W std::vector<VideoCaptureAPIs> getStreamBackends();
 
+/** @brief Returns list of available backends which works via `cv::VideoCapture(buffer)` */
+CV_EXPORTS_W std::vector<VideoCaptureAPIs> getBufferBackends();
```
Oh, seems like `getStreamBackends` is busy (see above)

opencv-alalek replied:
`getStreamBufferBackends` ?

dkurt replied:
Changed locally, can commit with https://github.com/opencv/opencv/pull/25584#discussion_r1896904920 if actual

opencv-alalek commented on modules/videoio/include/opencv2/videoio.hpp:
```diff
@@ -798,6 +798,14 @@ class CV_EXPORTS_W VideoCapture
+    The `params` parameter allows to specify extra parameters encoded as pairs `(paramId_1, paramValue_1, paramId_2, paramValue_2, ...)`.
+    See cv::VideoCaptureProperties
+    */
+    CV_WRAP explicit VideoCapture(Ptr<std::streambuf> source, int apiPreference, const std::vector<int>& params);
```
`const reference` here and below (or we should use `std::move()` massively)

opencv-alalek commented on modules/videoio/misc/python/pyopencv_videoio.hpp:
```diff
@@ -31,4 +31,87 @@ template<> bool pyopencv_to(PyObject* obj, cv::VideoCapture& stream, const ArgIn
+    PyObject* type = PyObject_GetAttrString(ioModule, "BufferedIOBase");
+    Py_DECREF(ioModule);
+    if (!PyObject_IsInstance(obj, type))
+        CV_Error(cv::Error::StsBadArg, "Input stream should be derived from io.BufferedIOBase");
```
`CV_Error` doesn't call cleanup (release GIL and release `type`).

```
if (...)
    has_error = true;

... cleanup ...

if (has_error)
    CV_Error(...);
```

opencv-alalek reviewed (commented):
Thank you for update!

There are some leftovers of `buffer` in the meaning of `stream`

`std::streambuf` is the I/O stream object at first (but "buffered" variant).
They are not buffers.

dkurt commented:
@opencv-alalek thanks! If this PR blocks plugin rebuild, may I ask you to commit changes to the branch? I’m out of the office before the end of the day.

dkurt commented:
> std::streambuf is the I/O stream object at first (but "buffered" variant).
They are not buffers.

Totally fine to keep stream notation everywhere. There is just a conflict in public API with filename (getStreamsBackends)

opencv-alalek commented on modules/videoio/src/cap_ffmpeg.cpp:
```diff
@@ -292,7 +314,43 @@ CvResult CV_API_CALL cv_capture_open_with_params(
+    try
+    {
+        cv::VideoCaptureParameters parameters(params, n_params);
+        cap = new CvCapture_FFMPEG_proxy(makePtr<ReadStreamPluginProvider>(opaque, read, seek), parameters);
```
> `makePtr<ReadStreamPluginProvider>(opaque, read, seek)`

Released right after this call because it is not saved anywhere.

dkurt replied:
As a workaround, `readStream = stream` for now

dkurt commented on modules/videoio/include/opencv2/videoio.hpp:
```diff
@@ -719,6 +719,25 @@ enum VideoCaptureOBSensorProperties{
 
+/** @brief Read data stream interface
+ */
+class CV_EXPORTS_W IReadStream
```
@opencv-alalek, @asmorkalov , need your opinion on naming as we ready to move it to the public API. There are `VideoCapture`, [IStreamSource ](https://docs.opencv.org/4.x/da/d03/classcv_1_1gapi_1_1wip_1_1IStreamSource.html), [VideoReader](https://docs.opencv.org/4.x/db/ded/classcv_1_1cudacodec_1_1VideoReader.html), [VideoWriter](https://docs.opencv.org/4.x/dd/d9e/classcv_1_1VideoWriter.html). So does it make sense to keep the order and name it `IStreamReader`? Moreover, it matches PyTorch's naming: https://docs.opencv.org/4.x/dd/d9e/classcv_1_1VideoWriter.html

opencv-alalek commented:
BTW, there is leftover with Python typing generator (map to `io.BufferedIOBase` or even `io.BufferedReader`)

opencv-alalek commented:
There is problem with iOS/Swift bindings:

```
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "Headers/opencv2.h"
        ^
/Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/lib/Release/opencv2.framework/Headers/opencv2.h:151:9: note: in file included from /Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/lib/Release/opencv2.framework/Headers/opencv2.h:151:
#import <opencv2/VideoCapture.h>
        ^
/Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/lib/Release/opencv2.framework/Headers/VideoCapture.h:288:1: error: duplicate declaration of method 'open:apiPreference:params:'
- (BOOL)open:(IStreamReader*)source apiPreference:(int)apiPreference params:(IntVector*)params NS_SWIFT_NAME(open(source:apiPreference:params:));
^
/Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/lib/Release/opencv2.framework/Headers/VideoCapture.h:224:1: note: previous declaration is here
- (BOOL)open:(NSString*)filename apiPreference:(int)apiPreference params:(IntVector*)params NS_SWIFT_NAME(open(filename:apiPreference:params:));
^
<unknown>:0: error: could not build Objective-C module 'opencv2'
```

Error message is not very clear to me

dkurt commented:
@opencv-alalek, Probably something like https://github.com/opencv/opencv/pull/25584/commits/0770b4907c13d058a9e8b1288cd45d278dbaf482 with

opencv-alalek commented:
Looks like fix works: https://github.com/opencv/opencv/actions/runs/12491953263/job/34858615536?pr=25584

@dkurt Feel free to push your changes.

dkurt commented:
@opencv-alalek no pending changes left

mshabunin commented on modules/videoio/include/opencv2/videoio.hpp:
```diff
@@ -852,6 +887,32 @@ class CV_EXPORTS_W VideoCapture
+
+    The method first calls VideoCapture::release to close the already opened file or camera.
+     */
+    virtual bool open(const Ptr<std::streambuf>& source, int apiPreference, const std::vector<int>& params);
```
Do we really need methods with `std::streambuf`? Maybe it would be better to make `StreambufReadStream` or similar class or just its factory public instead? For example:
```.cpp
shared_ptr<IStreamReader> source = createStdStream(filename); // can accept filename or &&filebuf or &&istream or &&ifstream
VideoCapture cap(source, ...);
```

dkurt replied:
Good point. But I think we can even remove it completely. `std::streambuf` might introduce more limitations than just seek/read interface.

mshabunin commented on modules/videoio/include/opencv2/videoio.hpp:
```diff
@@ -719,6 +719,25 @@ enum VideoCaptureOBSensorProperties{
+    /** @brief Sets the stream position
+     *
+     * @param offset Seek offset
+     * @param origin SEEK_SET / SEEK_END / SEEK_CUR
```
Wouldn't it be better to completely avoid C/C++ constants in public interface and introduce our own?

opencv-alalek replied:
FFmpeg (basically) and Python use them too (with same values).
IMHO, it is not necessary to introduce extra conversions to/from own values.

dkurt replied:
Agree, we can add such enum but potential user will still cast to integer or introduce long if/else/switch

mshabunin replied:
My concern was that user would be required to include `<stdio.h>`/`<cstdio>` in order to use this seemingly unrelated class. Though perhaps it's fine.
